File: GrandBy(grandby)/backend/app/schemas/diary.py
"""
다이어리 관련 Pydantic 스키마
"""

from pydantic import BaseModel
from datetime import date, datetime
from typing import Optional, List
from app.models.diary import AuthorType, DiaryStatus


class DiaryCreate(BaseModel):
    """다이어리 생성"""
    date: date
    title: Optional[str] = None
    content: str
    mood: Optional[str] = None
    status: DiaryStatus = DiaryStatus.PUBLISHED

class DiaryUpdate(BaseModel):
    """다이어리 수정"""
    title: Optional[str] = None
    content: Optional[str] = None
    mood: Optional[str] = None
    status: Optional[DiaryStatus] = None


class DiaryPhotoResponse(BaseModel):
    """다이어리 사진 응답"""
    photo_id: str
    photo_url: str
    created_at: datetime
    
    class Config:
        from_attributes = True


class DiaryResponse(BaseModel):
    """다이어리 응답"""
    diary_id: str
    user_id: str
    author_id: str
    author_name: Optional[str] = None
    date: date
    title: Optional[str] = None
    content: str
    mood: Optional[str] = None
    author_type: AuthorType
    is_auto_generated: bool
    status: DiaryStatus
    created_at: datetime
    updated_at: datetime
    comment_count: int = 0  # 댓글 개수
    photos: List[DiaryPhotoResponse] = []  # 사진 목록
    
    class Config:
        from_attributes = True


class DiaryCommentCreate(BaseModel):
    """댓글 생성"""
    content: str


class DiaryCommentResponse(BaseModel):
    """댓글 응답"""
    comment_id: str
    user_id: str
    content: str
    is_read: bool
    created_at: datetime
    user_name: str  # 유저 이름 추가
    user_role: str  # 유저 역할 추가
    
    class Config:
        from_attributes = True

File: GrandBy(grandby)/frontend/src/api/diary.ts
/**
 * 다이어리 API 클라이언트
 * 일기 CRUD 기능
 */

import apiClient from './client';

export interface DiaryPhoto {
  photo_id: string;
  photo_url: string;
  created_at: string;
}

export interface Diary {
  diary_id: string;
  user_id: string;
  author_id: string;
  author_name?: string;
  call_id?: string | null;  // 통화 ID 추가
  date: string; // YYYY-MM-DD 형식
  title?: string | null;
  content: string;
  mood?: string | null; // happy, sad, calm, excited, angry 등
  author_type: 'elderly' | 'caregiver' | 'ai';
  is_auto_generated: boolean;
  status: 'draft' | 'published';
  created_at: string;
  updated_at: string;
  comment_count?: number; // 댓글 개수
  photos?: DiaryPhoto[]; // 사진 목록
}

export interface DiaryCreate {
  date: string; // YYYY-MM-DD 형식
  title?: string;
  content: string;
  mood?: string;
  status?: 'draft' | 'published';
}

export interface DiaryUpdate {
  title?: string;
  content?: string;
  mood?: string;
  status?: 'draft' | 'published';
}

/**
 * 다이어리 목록 조회
 * 
 * @param params - 조회 옵션
 * @returns 다이어리 목록
 */
export const getDiaries = async (params?: {
  skip?: number;
  limit?: number;
  start_date?: string;
  end_date?: string;
  elderly_id?: string;
}): Promise<Diary[]> => {
  const response = await apiClient.get<Diary[]>('/api/diaries/', { params });
  return response.data;
};

/**
 * 다이어리 상세 조회
 * 
 * @param diaryId - 다이어리 ID
 * @returns 다이어리 상세 정보
 */
export const getDiary = async (diaryId: string): Promise<Diary> => {
  const response = await apiClient.get<Diary>(`/api/diaries/${diaryId}`);
  return response.data;
};

/**
 * 다이어리 작성
 * 
 * @param data - 다이어리 작성 데이터
 * @returns 생성된 다이어리 (보호자인 경우 여러 개)
 */
export const createDiary = async (data: DiaryCreate): Promise<Diary[]> => {
  const response = await apiClient.post<Diary[]>('/api/diaries/', data);
  return response.data;
};

/**
 * 다이어리 수정
 * 
 * @param diaryId - 다이어리 ID
 * @param data - 수정 데이터
 * @returns 수정된 다이어리
 */
export const updateDiary = async (diaryId: string, data: DiaryUpdate): Promise<Diary> => {
  const response = await apiClient.put<Diary>(`/api/diaries/${diaryId}`, data);
  return response.data;
};

/**
 * 다이어리 삭제
 * 
 * @param diaryId - 다이어리 ID
 * @returns 삭제 결과 메시지
 */
export const deleteDiary = async (diaryId: string): Promise<{ message: string }> => {
  const response = await apiClient.delete<{ message: string }>(`/api/diaries/${diaryId}`);
  return response.data;
};

// ==================== 댓글 API ====================

export interface DiaryComment {
  comment_id: string;
  user_id: string;
  content: string;
  is_read: boolean;
  created_at: string;
  user_name: string;
  user_role: string;
}

export interface CommentCreateRequest {
  content: string;
}

/**
 * 댓글 목록 조회
 * 
 * @param diaryId - 다이어리 ID
 * @returns 댓글 목록
 */
export const getComments = async (diaryId: string): Promise<DiaryComment[]> => {
  const response = await apiClient.get<DiaryComment[]>(`/api/diaries/${diaryId}/comments`);
  return response.data;
};

/**
 * 댓글 작성
 * 
 * @param diaryId - 다이어리 ID
 * @param data - 댓글 내용
 * @returns 생성된 댓글
 */
export const createComment = async (diaryId: string, data: CommentCreateRequest): Promise<DiaryComment> => {
  const response = await apiClient.post<DiaryComment>(`/api/diaries/${diaryId}/comments`, data);
  return response.data;
};

/**
 * 댓글 삭제
 * 
 * @param diaryId - 다이어리 ID
 * @param commentId - 댓글 ID
 * @returns 삭제 결과 메시지
 */
export const deleteComment = async (diaryId: string, commentId: string): Promise<{ message: string }> => {
  const response = await apiClient.delete<{ message: string }>(`/api/diaries/${diaryId}/comments/${commentId}`);
  return response.data;
};

